Openrouter returns 524 when querying models

happened twice now

```sh
2024/06/13 11:10:12 Checking provider "openrouter" for models
2024/06/13 11:11:52 ERROR: could not query models for provider "openrouter": error, status code: 524, message: invalid character '<' looking for beginning of value
panic: ERROR: could not query models for provider "openrouter": error, status code: 524, message: invalid character '<' looking for beginning of value

goroutine 1 [running]:
log.(*Logger).Panicf(0x98a81f?, {0x9a14d4?, 0xc0002a8400?}, {0xc00031f1e8?, 0x0?, 0x13?})
        /home/simon/symflower/.devenv/go@1.21.5/go/src/log/log.go:310 +0x6e
github.com/symflower/eval-dev-quality/cmd/eval-dev-quality/cmd.(*Evaluate).Execute(0xc0002d2000, {0x7fad9c03f8f8?, 0xc0002d2000?, 0x914180?})
        /home/simon/repos/eval-dev-quality/cmd/eval-dev-quality/cmd/evaluate.go:243 +0x1da5
github.com/symflower/eval-dev-quality/cmd/eval-dev-quality/cmd.Execute.func1({0x7fad9c03f8f8, 0xc0002d2000}, {0xc0002e2000, 0x0, 0xb})
        /home/simon/repos/eval-dev-quality/cmd/eval-dev-quality/cmd/command.go:34 +0x8c
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc0002c9730, {0xc00012e010?, 0xb, 0xb})
        /home/simon/symflower/pkg/mod/github.com/jessevdk/go-flags@v1.5.1-0.20210607101731-3927b71304df/parser.go:333 +0x954
github.com/symflower/eval-dev-quality/cmd/eval-dev-quality/cmd.Execute(0xc0002998f0, {0xc00012e010, 0xb, 0xb})
        /home/simon/repos/eval-dev-quality/cmd/eval-dev-quality/cmd/command.go:37 +0x247
main.main()
        /home/simon/repos/eval-dev-quality/cmd/eval-dev-quality/main.go:11 +0x65

```

--- a/provider/openrouter/openrouter.go
+++ b/provider/openrouter/openrouter.go
@@ -4,7 +4,9 @@ import (
 	"context"
 	"errors"
 	"strings"
+	"time"
 
+	"github.com/avast/retry-go"
 	pkgerrors "github.com/pkg/errors"
 	"github.com/sashabaranov/go-openai"
 
@@ -52,13 +54,28 @@ func (p *Provider) ID() (id string) {
 // Models returns which models are available to be queried via this provider.
 func (p *Provider) Models() (models []model.Model, err error) {
 	client := p.client()
-	ms, err := client.ListModels(context.Background())
-	if err != nil {
-		return nil, pkgerrors.WithStack(err)
+
+	var responseModels openai.ModelsList
+	if err := retry.Do( // Query available models with a retry logic cause "openrouter.ai" has failed us in the past.
+		func() error {
+			ms, err := client.ListModels(context.Background())
+			if err != nil {
+				return pkgerrors.WithStack(err)
+			}
+			responseModels = ms
+
+			return nil
+		},
+		retry.Attempts(3),
+		retry.Delay(5*time.Second),
+		retry.DelayType(retry.BackOffDelay),
+		retry.LastErrorOnly(true),
+	); err != nil {
+		return nil, err
 	}
 
-	models = make([]model.Model, len(ms.Models))
-	for i, model := range ms.Models {
+	models = make([]model.Model, len(responseModels.Models))
+	for i, model := range responseModels.Models {
 		models[i] = llm.NewModel(p, p.ID()+provider.ProviderModelSeparator+model.ID)
 	}
 

